### PR TITLE
Implement Linux machine fingerprints

### PR DESCRIPTION
**Pull request by daleglass**
We do this by using either /etc/machine-id, or hashing the MACs of all
non-virtual networking devices.

We make this fingerpint unique to our platform by adding our own string
to the hash, and add some stretching for good measure. This way Vircadia's
fingerprint isn't going to match the system's ID or anything another
application generates.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe-_9f1581a9.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe-.tar_9f1581a9.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe-.tar_9f1581a9.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe-_9f1581a9.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe-_9f1581a9.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-Quest-Alpha-PR1092-41370fe_9f1581a9.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-41370fe_9f1581a9.apk

File: libraries/networking/src/FingerprintUtils.cpp [hunk 1]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
 #ifdef Q_OS_LINUX
-    // sadly need to be root to get smbios guid from linux, so 
-    // for now lets do nothing.
+    // As per the documentation:
+    // https://man7.org/linux/man-pages/man5/machine-id.5.html
+    //
+    // we use the machine id as a base, but add an application-specific key to it.
+    // If machine-id isn't available, we try hardware networking devices instead.
+
+    QCryptographicHash hash{QCryptographicHash::Keccak_256};
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:48]:
```suggestion
    QCryptographicHash hash(QCryptographicHash::Keccak_256);
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 2]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+
+    QFile machineIdFile{"/etc/machine-id"};
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:50]:
```suggestion
    QFile machineIdFile("/etc/machine-id");
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 3]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+    if (!machineIdFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        // No machine ID, probably no systemd.
+        qCWarning(networking) << "Failed to open /etc/machine-id";
+
+        QDir netDevicesDir{"/sys/class/net"};
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:55]:
```suggestion
        QDir netDevicesDir("/sys/class/net");
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 4]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+        QFileInfoList netDevicesInfo = netDevicesDir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name);
+
+        if ( netDevicesInfo.empty() ) {
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:58]:
```suggestion
        if (netDevicesInfo.empty()) {
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 5]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+            // Let getMachineFingerprint handle this contingency
+            qCWarning(networking) << "Failed to find any hardware networking devices";
+            return "";
+        }
+
+        for( auto& fileInfo : netDevicesInfo ) {
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:64]:
```suggestion
        for (auto& fileInfo : netDevicesInfo) {
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 6]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+    } else {
+        QByteArray data = machineIdFile.readAll();
+        hash.addData(data);
+    }
+
+    // Makes this hash unique to us
+    hash.addData("Vircadia");
+
+    // Stretching
+    for(int i=0;i<65535;i++) {
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:91]:
```suggestion
    for (int i=0; i<65535; i++) {
```

File: libraries/networking/src/FingerprintUtils.cpp [hunk 7]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+        hash.addData(hash.result());
+    }
+
+    QByteArray result = hash.result();
+    result.resize(128/8); // GUIDs are 128 bit numbers
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:96]:
```suggestion
    result.resize(128 / 8); // GUIDs are 128 bit numbers
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331-_5345ba27.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331-.tar_5345ba27.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331-.tar_5345ba27.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331-_5345ba27.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331-_5345ba27.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-Quest-Alpha-PR1092-fd08331_5345ba27.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-fd08331_5345ba27.apk

File: libraries/networking/src/FingerprintUtils.cpp [hunk 8]
```diff
@@ -36,8 +39,80 @@ QUuid FingerprintUtils::_machineFingerprint { QUuid() };
...
+    } else {
+        QByteArray data = machineIdFile.readAll();
+        hash.addData(data);
+    }
+
+    // Makes this hash unique to us
+    hash.addData("Vircadia");
+
+    // Stretching
+    for (int i=0; i<65535; i++) {
```
Comment (ctrlaltdavid) [inline libraries/networking/src/FingerprintUtils.cpp:91]:
```suggestion
    for (int i=0; i < 65535; i++) {
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1-_1ed767df.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1-.tar_1ed767df.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1-.tar_1ed767df.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1-_1ed767df.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1_1ed767df.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-Quest-Alpha-PR1092-91df8b1_1ed767df.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1092/Vircadia-PR1092-91df8b1-_1ed767df.exe

**Comment by digisomni** [comment]
To test: Please run Interface multiple times and ensure your fingerprint stays the same.
You can check your fingerprint with the domain-server panel, or search the log.

**Comment by JulianGro** [comment]
On my machine where `/etc/machine-id` is a things, the fingerprint stays the same.

**Pull request merged**